Add configmap for prometheus.yml config (and use upstream prometheus docker image)

stephen-soltesz opened the pull request:
This PR removes the need to create a custom prometheus docker image by adding a ConfigMap configuration for the prometheus configuration.

Using the ConfigMap, operators can update a configuration file and restart the pod.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/prometheus-support/7)
<!-- Reviewable:end -->

salarcon215 commented:
I'm re-running these steps. They seem accurate so far. However, a nagging thought that I keep coming back to is that what we're missing here is a very high-level, very simple, plain English blurb about the intended outcome of each step. ("Step" here means items with H1 and H2, or single and double hash marks.) 

For example, the section called "Testing: Using explicit kubectl commands" says "This option is helpful for testing. There is no persistent storage."  Cool, but what are would we be testing? How would we notice success or failure? In what situation would that bare bones testing scenario be helpful?

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[README.md, line 73 at r2](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfnNiNKAkj7g_WmWMZd:-KfnNiNLvuOazbFvTXcI:b-stf4hq) ([raw file](https://github.com/m-lab/prometheus-support/blob/4393ba4cffe9f44dc5a936e12cf43cc22cf2d1b6/README.md#L73)):*
> ```Markdown
> 
> [file_sd_config]: https://prometheus.io/docs/operating/configuration/#file_sd_config
> ```

I like very much the inclusion of the URL to see the effect of the change. But is there also a kubectl command we can use to see the addition of the targets directory?

---

*[README.md, line 130 at r2](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfnO0Sxfs1-AM0n9gZh:-KfnO0Sxfs1-AM0n9gZi:b9ke8d) ([raw file](https://github.com/m-lab/prometheus-support/blob/4393ba4cffe9f44dc5a936e12cf43cc22cf2d1b6/README.md#L130)):*
> ```Markdown
> Note: Configmaps only support text data. Secrets may be an alternative for
> binary data. https://github.com/kubernetes/kubernetes/issues/32432
> ```

Similarly, here, what can we use to make sure the mount works? I think maybe we want  `kubectl get pods` and/or 
```podname=$( kubectl get pods -o name --selector='run=prometheus-server' )
kubectl describe pod $podname```
Does that seem right?
```


---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-:-KfnMZBMZWYOx3EywnFa:bhmbbnk)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Okay. I put some more effort into clarifying the README.

I've removed the "testing" section because that kind of doesn't make sense now that we're using a configmap for configuration and `kubectl run` doesn't allow volume mounts from the command line.

I've moved the legacy targets section to follow creating the configmaps and deployments. The sections should roughly be in dependency order.

How does it look?

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[README.md, line 73 at r2](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfnNiNKAkj7g_WmWMZd:-KfrWSgpO1MaEkj-h3W8:b-3jnajh) ([raw file](https://github.com/m-lab/prometheus-support/blob/4393ba4cffe9f44dc5a936e12cf43cc22cf2d1b6/README.md#L73)):*
<details><summary><i>Previously, salarcon215 wrote…</i></summary><blockquote>

I like very much the inclusion of the URL to see the effect of the change. But is there also a kubectl command we can use to see the addition of the targets directory?
</blockquote></details>

I've added a kubectl command to list the `legacy-targets` directory.  How does it look?

---

*[README.md, line 130 at r2](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfnO0Sxfs1-AM0n9gZh:-KfrYVWoi9EpUmd6wx4t:b-jgw1em) ([raw file](https://github.com/m-lab/prometheus-support/blob/4393ba4cffe9f44dc5a936e12cf43cc22cf2d1b6/README.md#L130)):*
<details><summary><i>Previously, salarcon215 wrote…</i></summary><blockquote>

Similarly, here, what can we use to make sure the mount works? I think maybe we want  `kubectl get pods` and/or 
```podname=$( kubectl get pods -o name --selector='run=prometheus-server' )
kubectl describe pod $podname```
Does that seem right?
</blockquote></details>

Good question. I've added an additional subsection for verifying that a ConfigMap is moutned. How does it look?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-:-Kfrk9tR1L07bWrkZ0e9:b30yuxe)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Thank you for the careful feedback. You caught several bugs! :-D

PTAL?

---

Review status: 0 of 3 files reviewed at latest revision, 9 unresolved discussions.

---

*[README.md, line 20 at r3](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-Kfx902WQUViZgOixMgz:-KfxHzU8jGm-cblqMHP2:b70di6j) ([raw file](https://github.com/m-lab/prometheus-support/blob/89df580cee50abc6cdbc5c4e62d98ac3f067ae35/README.md#L20)):*
<details><summary><i>Previously, salarcon215 wrote…</i></summary><blockquote>

Is there an extra `(` in front of "yet"?
</blockquote></details>

Oops! This is wrong twice. There was a missing `)` at the end. But the markdown href syntax is incorrect also. Fixed.

---

*[README.md, line 42 at r3](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfxAJ1QrD5mQ-cWnIMf:-KfxIio9etPUetrh7wLW:bt2vr1b) ([raw file](https://github.com/m-lab/prometheus-support/blob/89df580cee50abc6cdbc5c4e62d98ac3f067ae35/README.md#L42)):*
<details><summary><i>Previously, salarcon215 wrote…</i></summary><blockquote>

My inner copy editor says this "while" gives the impression of carrying some special loop-like significance. So I recommend something like: "Although the flag is called `--from-file`, it accepts a directory. It creates a ConfigMap with keys equal to the filenames, and values equal to the content of the file."
</blockquote></details>

Fixed.

---

*[README.md, line 57 at r3](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfxBUTwBR6EXvQdtHJT:-KfxIpQBds2nWTazRp6I:b-d0klhz) ([raw file](https://github.com/m-lab/prometheus-support/blob/89df580cee50abc6cdbc5c4e62d98ac3f067ae35/README.md#L57)):*
<details><summary><i>Previously, salarcon215 wrote…</i></summary><blockquote>

Did you mean "configuration"?
</blockquote></details>

:-) Fixed.

---

*[README.md, line 129 at r3](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfxDGQ_z2nbgvNPIyce:-KfxJL8fZTBaLPaGWov3:b-z9wueq) ([raw file](https://github.com/m-lab/prometheus-support/blob/89df580cee50abc6cdbc5c4e62d98ac3f067ae35/README.md#L129)):*
<details><summary><i>Previously, salarcon215 wrote…</i></summary><blockquote>

Is the latest container tag specified here?

```
    spec:                                                                      
      containers:                                                              
      - image: prom/prometheus:v1.5.2                                          
        # Note: this value appears to be ignored and the actual pod name is    
        # derived from the Deployment.metadata.name. However, removing this    
        # value results in a configuration error.
```
I think this is worth clarifying. IIRC, the Docker Way to specify latest is to terminate the Dockerhub path with `:latest`. This bit of prometheus.yml looks like a tag spec, but the comment makes it seem like the version will be ignored. Also, how will the user make sure which tag level is the latest? Will they look it up in Dockerhub or can it be specified using "latest" or similar?
</blockquote></details>

I've changed the phrasing to "to reference the *current* stable prometheus container tag" to avoid confusion with "latest". We don't actually want the latest tag because it can change over time. We want a stable tag that won't change. 

I've also clarified the comment in `prometheus.yml`. The comment referred to the `name` attribute.

---

*[README.md, line 141 at r3](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfxEdy-MVfDR1OnUxW_:-KfxKK0b9QHa4ydwK7GV:b-89q77x) ([raw file](https://github.com/m-lab/prometheus-support/blob/89df580cee50abc6cdbc5c4e62d98ac3f067ae35/README.md#L141)):*
<details><summary><i>Previously, salarcon215 wrote…</i></summary><blockquote>

Excellent context, helpful for building a mental model of which things change under what conditions.
</blockquote></details>

Cool!

---

*[README.md, line 192 at r3](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfxF9HLgmSpMEZ0NAaX:-KfxKM7pIMpXR2aQNPcd:btojbfd) ([raw file](https://github.com/m-lab/prometheus-support/blob/89df580cee50abc6cdbc5c4e62d98ac3f067ae35/README.md#L192)):*
<details><summary><i>Previously, salarcon215 wrote…</i></summary><blockquote>

These examples are a great addition.
</blockquote></details>

Great!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-:-KfxKnNrhzKABIwglSbd:bqfk5vb)*
<!-- Sent from Reviewable.io -->

salarcon215 commented:
A couple more small nits to sweep up. You are welcome to bounce it back for final thumbs up, but I'm granting a Friday afternoon implicit trust LGTM so you can clear your queue for next week if you like!

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[README.md, line 42 at r3](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfxAJ1QrD5mQ-cWnIMf:-Kg-dgVAikOdaEhG_sHK:b3pz2z0) ([raw file](https://github.com/m-lab/prometheus-support/blob/89df580cee50abc6cdbc5c4e62d98ac3f067ae35/README.md#L42)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Fixed.
</blockquote></details>

"...the kubernetes creates"...  Does The Soltesz mean that the kubernetes cluster or command creates?

---

*[README.md, line 129 at r3](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfxDGQ_z2nbgvNPIyce:-Kg-eKl6tzFL5MJpHOYo:b-jzjjhe) ([raw file](https://github.com/m-lab/prometheus-support/blob/89df580cee50abc6cdbc5c4e62d98ac3f067ae35/README.md#L129)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I've changed the phrasing to "to reference the *current* stable prometheus container tag" to avoid confusion with "latest". We don't actually want the latest tag because it can change over time. We want a stable tag that won't change. 

I've also clarified the comment in `prometheus.yml`. The comment referred to the `name` attribute.
</blockquote></details>

Ok, cool. But how do we make sure we're using the current stable version? Do we look at https://hub.docker.com/r/prom/prometheus/tags/ and choose the highest numbered (as opposed to named, like "latest") version?

---

*[README.md, line 33 at r4](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-Kg-afxInSWgOYpQVNMu:-Kg-afxJZ2savNTB1M2g:bk9z8j2) ([raw file](https://github.com/m-lab/prometheus-support/blob/c7d3fe06cb1d6167a8090520189edfa3d1538b3d/README.md#L33)):*
> ```Markdown
> ## ConfigMaps
> 
> Many services like prometheus provide canonical docker images publised to
> ```

"published"?

---

*[README.md, line 34 at r4](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-Kg-b1Bmd5PUODbGtjib:-Kg-b1BnIfNtFPtDIq9I:b-hoeesx) ([raw file](https://github.com/m-lab/prometheus-support/blob/c7d3fe06cb1d6167a8090520189edfa3d1538b3d/README.md#L34)):*
> ```Markdown
> 
> Many services like prometheus provide canonical docker images publised to
> dockerhub (or other registry). We can customize the deployment by changing the
> ```

Capitalize "Dockerhub", include link. Is there an M-Lab organization on Dockerhub yet? I see Greg's experiments but not our own. totally fine if we don't have (or need) one, but if we have it, let's link to it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-:-Kg-fJZqXecam7ikkedf:bwd4jvg)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Thanks for the review! I'll use the LGTM now. We can clean up any remaining issues with subsequent PRs.

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[README.md, line 42 at r3](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfxAJ1QrD5mQ-cWnIMf:-Kg-kWQpTrHbadg_XkZP:b-9dx9f4) ([raw file](https://github.com/m-lab/prometheus-support/blob/89df580cee50abc6cdbc5c4e62d98ac3f067ae35/README.md#L42)):*
<details><summary><i>Previously, salarcon215 wrote…</i></summary><blockquote>

"...the kubernetes creates"...  Does The Soltesz mean that the kubernetes cluster or command creates?
</blockquote></details>

How about just "kubernetes creates"?

---

*[README.md, line 129 at r3](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-KfxDGQ_z2nbgvNPIyce:-Kg-lgLE1TTOaamrP8MW:b-k0wvfg) ([raw file](https://github.com/m-lab/prometheus-support/blob/89df580cee50abc6cdbc5c4e62d98ac3f067ae35/README.md#L129)):*
<details><summary><i>Previously, salarcon215 wrote…</i></summary><blockquote>

Ok, cool. But how do we make sure we're using the current stable version? Do we look at https://hub.docker.com/r/prom/prometheus/tags/ and choose the highest numbered (as opposed to named, like "latest") version?
</blockquote></details>

Yes, exactly. Added a link to the k8s/prometheus.yml file.

---

*[README.md, line 33 at r4](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-Kg-afxInSWgOYpQVNMu:-Kg-jkMQllXQPidToP2v:bt2vr1b) ([raw file](https://github.com/m-lab/prometheus-support/blob/c7d3fe06cb1d6167a8090520189edfa3d1538b3d/README.md#L33)):*
<details><summary><i>Previously, salarcon215 wrote…</i></summary><blockquote>

"published"?
</blockquote></details>

Fixed.

---

*[README.md, line 34 at r4](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-Kg-b1Bmd5PUODbGtjib:-Kg-kEwyQiNLmowxhAJU:b4pgv9i) ([raw file](https://github.com/m-lab/prometheus-support/blob/c7d3fe06cb1d6167a8090520189edfa3d1538b3d/README.md#L34)):*
<details><summary><i>Previously, salarcon215 wrote…</i></summary><blockquote>

Capitalize "Dockerhub", include link. Is there an M-Lab organization on Dockerhub yet? I see Greg's experiments but not our own. totally fine if we don't have (or need) one, but if we have it, let's link to it.
</blockquote></details>

I've included a link to the prometheus dockerhub page, since that's the most important one in this context.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-support/7#-:-Kg-mJQLIXZIZwhaeWiS:b-dp2jm5)*
<!-- Sent from Reviewable.io -->
